Add convenience function to get ply-wise data

janvonrickenbach opened the pull request:
Add convenience function to get filtered ply wise data.

janvonrickenbach commented on src/ansys/dpf/composites/layup_info/_layup_info.py:
```diff
@@ -219,8 +219,8 @@ def get_dpf_material_id_by_analyis_ply_map(
         DPF data source with rst file or streams_provider. The streams provider is
         available from :attr:`.CompositeModel.core_model` (under metadata.streams_provider).
 
```
Fixes error in documetation creation

PipKat commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+
+
+class ReductionStrategy(Enum):
+    """The reduction strategy to get from spot values (BOT, MID, TOP) to a single value."""
```
```suggestion
    """Provides the reduction strategy to get from spot values (BOT, MID, TOP) to a single value."""
```

PipKat commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+"""Methods to get ply wise data from a result field."""
```
```suggestion
"""Methods to get ply-wise data from a result field."""
```

PipKat commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+    field:
+        The field to extract data from.
+    ply_name:
+        The name of the ply to extract data from.
+    mesh:
+        The meshed region. Needs to be enriched with composite information.
+        Use CompositeModel.get_mesh() to get the meshed region.
+    reduction_strategy:
+        The reduction strategy to get from spot values (BOT, MID, TOP) to a single value
+        per corner node and layer. The default is AVG.
+    requested_location:
+        The location of the output field. The default is elemental_nodal. Supported are
+        elemental_nodal, elemental, and nodal.
+    component:
+        The component to extract data from. Can be an int or an IntEnum. The default is 0.
```
```suggestion
        Name of the ply to extract data from.
    mesh :
        Meshed region. Needs to be enriched with composite information.
        Use the ``CompositeModel.get_mesh()`` method to get the meshed region.
    reduction_strategy :
        Reduction strategy for getting from spot values (BOT, MID, TOP) to a single value
        per corner node and layer. The default is ``AVG``.
    requested_location :
        Location of the output field. The default is ``"elemental_nodal"``. Options are
        ``"elemental"``, ``"elemental_nodal"``, and ``"nodal"``.
    component : int or IntEnum, optional
        Component to extract data from. The default is ``0``.
```
List default first and then provide a list of all options in alphabetical order after.

PipKat commented on tests/ply_wise_data_test.py:
```diff
@@ -0,0 +1,140 @@
+]
+
+
+def get_all_spot_values_first_element_first_node(stress_field, component=0):
```
I see that a lot of these methods don't have docstrings yet.

janvonrickenbach replied:
Hi @PipKat Thanks for the review. These function are just helper functions for tests and are not part of the public API. Therefore they don't need documentation.

greschd commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+    Parameters
+    ----------
+    field:
+        The field to extract data from.
```
Hi @PipKat, thanks for the review!

I think (will need to double-check) this repository is set up s.t. the types are automatically added from the type hints (see e.g. lines 24-29) into the rendered documentation. 
This means we only need to add them in one place, and they are automatically checked and inserted into the rendered documentation.

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -55,16 +58,52 @@
+
+
+# %%
+# Generic client side filtering
```
```suggestion
# Generic client-side filtering
```
Here, "client-side" is a compound adjective. The hyphen is required.

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -82,23 +121,45 @@
-# List all available analysis plies.
-all_ply_names = get_all_analysis_ply_names(composite_model.get_mesh())
-all_ply_names
+# Filtering by material
```
```suggestion
# Filter by material
```

greschd reviewed (approved):
LGTM, just a typo and one minor comment.

janvonrickenbach commented on examples/006_filter_composite_data_example.py:
```diff
@@ -55,16 +58,52 @@
+# or maximum/minimum over the spot locations.
+# It also supports selecting a specific spot (TOP, MID, BOT) directly.
+# In this example we select the maximum value over all spots for each node and then request
+# the elemental location which implies averaging over all nodes in an element.
```
It is a plain average. I clarified in the doc string of the function and renamed the argument.

greschd commented:
> It is a plain average. I clarified in the doc string of the function and renamed the argument.

Could it be this change was made in `feat/fatigue_example` instead of this branch?

janvonrickenbach commented:
> > It is a plain average. I clarified in the doc string of the function and renamed the argument.
> 
> Could it be this change was accidentally made in `feat/fatigue_example` instead of this branch?

@greschd Yes, I made it there to avoid possible merge conflicts since I added also the comment concerning averaging in the fatigue branch. Sorry for the confunsion and thanks for changing the base branch on the fatigue PR.

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -4,9 +4,15 @@
 layered composites. You can filter strains and stresses by material, layer, or
-analysis ply. The example filters data by layer, spot, and node, as well as material
+analysis ply. Filtering by analysis ply is implemented on the server side and
+exposed with the function :func:`.get_ply_wise_data`. In this case the data is
```
```suggestion
exposed with the :func:`.get_ply_wise_data` function. In this case, the data is
```
Goggle dev style guide says that the noun should follow (not precede) the code entity for improved readability and comprehension.

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -55,16 +58,52 @@
+all_ply_names
+
+# %%
+# The easiest way to filter data by analysis ply is to use the function :func:`.get_ply_wise_data`.
```
```suggestion
# The easiest way to filter data by analysis ply is to use the :func:`.get_ply_wise_data` function.
```

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -55,16 +58,52 @@
+
+# %%
+# The easiest way to filter data by analysis ply is to use the function :func:`.get_ply_wise_data`.
+# This function supports different reduction strategies such as computing the average
+# or maximum/minimum over the spot locations.
```
```suggestion
# This function supports different reduction strategies such as computing the average,
# maximum, or minimum over the spot locations.
```
We should avoid using ``/" as the reader doesn't know whether it means "and" or "or".

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -55,16 +58,52 @@
+# This function supports different reduction strategies such as computing the average
+# or maximum/minimum over the spot locations.
+# It also supports selecting a specific spot (TOP, MID, BOT) directly.
+# In this example we select the maximum value over all spots for each node and then request
+# the elemental location which implies averaging over all nodes in an element.
```
```suggestion
# This example selects the maximum value over all spots for each node and then requests
# the elemental location, which implies averaging over all nodes in an element.
```
Avoid "we" and either use "you" or make another noun the subject, as I did here by making "example" the subject.

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -55,16 +58,52 @@
+# It also supports selecting a specific spot (TOP, MID, BOT) directly.
+# In this example we select the maximum value over all spots for each node and then request
+# the elemental location which implies averaging over all nodes in an element.
+# Using :func:`.get_ply_wise_data` has the advantage that all the averaging and filtering
+# is done on the server side.
```
```suggestion
# Using the :func:`.get_ply_wise_data` function has the advantage that all the averaging
#  and filtering is done on the server side.
```

PipKat commented on examples/006_filter_composite_data_example.py:
```diff
@@ -55,16 +58,52 @@
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# This example shows how to filter data by layer, spot, and node using the generic filtering on
+# the client side.
+# Here we plot stress values in the material direction for the first node and top spot.
```
```suggestion
# This code plots stress values in the material direction for the first node and top spot.
```

PipKat commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+
+
+class ReductionStrategy(Enum):
+    """The reduction strategy to get from spot values (BOT, MID, TOP) to a single value."""
```
```suggestion
    """Provides the reduction strategy for getting from spot values (BOT, MID, TOP) to a single value."""
```
A description for a class should always being in a verb that ends in "s" or "es" so that summary tables are consistent in format.

PipKat commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+    ply_name:
+        Name of the ply to extract data from.
+    mesh :
+        Meshed region. Needs to be enriched with composite information.
```
```suggestion
        Meshed region to enrich with composite information.
```
I redid this because "Needs to be enriched...." isn't a complete sentence on its own. Please revise if this isn't what was meant.

janvonrickenbach commented on examples/006_filter_composite_data_example.py:
```diff
@@ -116,31 +177,3 @@
-        element_info = composite_model.get_element_info(element_id)
-        assert element_info is not None
-
-        selected_indices = get_selected_indices_by_dpf_material_ids(element_info, [ud_material_id])
```
The example is still there (line 126). I just moved the old selection by analysis ply to the end because I consider it the most advanced example.

janvonrickenbach commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+    mesh :
+        Meshed region enriched with composite information.
+        Use the ``CompositeModel.get_mesh()`` method to get the meshed region.
+    reduction_strategy :
```
Yes. It is documented by the enum type. I would consider this enough but I can also state the supported mode explicitly.

janvonrickenbach commented on src/ansys/dpf/composites/ply_wise_data.py:
```diff
@@ -0,0 +1,87 @@
+__all__ = ("ReductionStrategy", "get_ply_wise_data")
+
+
+class ReductionStrategy(Enum):
```
We could move it to constants. But personally I prefer to keep enums somewhat local as long as there is no usage across modules.
